[JIT] enable profiling optional tensors

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70532

This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

Differential Revision: [D33714748](https://our.internmc.facebook.com/intern/diff/D33714748)

davidberard98 committed:
[WIP][JIT] enable profiling optional tensors

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] enable profiling optional tensors"




[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] enable profiling optional tensors"




[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] enable profiling optional tensors"




[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -200,21 +203,39 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
+      }
+
+      if (seen_none || input_is_optional) {
+        it->input()->setType(OptionalType::create(merged_tensor_type));
       } else {
-        it->input()->setType(input_type->merge(*profiled_type));
+        it->input()->setType(merged_tensor_type);
       }
```
@Krovatkin how should this behave?

1) an Optional[Tensor] always gets specialized to `Optional[specialized-tensor]`, e.g. if profiler sees "tensor(2, 2), tensor(2, 3), and tensor(2, 4)", then it'll be `Optional[Float(2, ?)]`
2) Optional[Tensor] can get specialized to `specialized-tensor` if the profiler never sees `None`, or `Optional[specialized-tensor]` if the profiler does see `None`

Krovatkin replied:
@davidberard98 I'd suggest 1. or to even be on the safe side to always set the type to `Optional::of(Tensor::get)` as if we never even profiled the optional tensor types. I'd like to keep this PR focused on *profiling* optional tensors rather than *extending optimizations to consume this new profiled type* and option 2 seems to require cascading changes in TE and possibly in NVFuser.

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33714748).

davidberard98 commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -201,21 +214,26 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
       // type refinement node so uses can have the correct corresponding type.
-      if (profiled_type == TensorType::get()) {
+      if (profiled_type == TensorType::get() && !input_is_optional) {
         continue;
```
@Krovatkin curious, shouldn't this also have `it.destroyCurrent()` ? Or is there a DCE pass that will remove the profile node?

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

Differential Revision: [D33714748](https://our.internmc.facebook.com/intern/diff/D33714748)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33714748).

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

Differential Revision: [D33714748](https://our.internmc.facebook.com/intern/diff/D33714748)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33714748).

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -201,21 +214,26 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
         continue;
       }
+
+      if (input_is_optional) {
```
nit: I wonder if we could pull this before `if (profiled_type == TensorType::get() && !input_is_optional)` so we don't have to check for `!input_is_optional`

This way the intent to ignore profiling information from optional tensors might be a bit clearer.

Krovatkin reviewed (approved):
:shipit:

davidberard98 commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -1992,6 +1992,66 @@ TEST(ProfilerTest, Basic) {
   checkShape(tanh_n->inputs().at(0)->node()->ty(attr::profiled_type), eltwise);
 }
 
+TEST(ProfilerTest, OptionalProfiling) {
```
@Krovatkin  I think this is actually testing the case where sometimes the tensor is provided, and sometimes it is `None` (which is new logic, since before we didn't have optional tensors

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

Differential Revision: [D33714748](https://our.internmc.facebook.com/intern/diff/D33714748)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] enable profiling optional tensors"


This adds profiling to Optional[Tensor] types

First, in profiling_record.cpp, profiling nodes are added to Optional[Tensor] inputs. The nodes record
(a) whether or not any `None` types are encountered, and
(b) of the Tensor types, what's the most specific type matching all of non-null tensors that were encoutered (shape, dtype, etc.)

In tensorexpr_fuser, when specializing types based on the profiled information, an Optional[Tensor] type will always be Optional[], but the Tensor type contained in the optional type can be specialized (e.g. `Optional[Float(2x2x2, cpu, etc)]`)

Differential Revision: [D33714748](https://our.internmc.facebook.com/intern/diff/D33714748)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33714748).
